kernel_gateway/services/kernels/pool.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from tornado.locks import Semaphore
+from tornado import gen
+
+
+class KernelPool():
+    '''
+    A class to maintain a pool of kernel and control access to the individual kernels.
+    Kernels are protected by a borrower/lender pattern.
+    '''
+    def __init__(self, prespawn_count, kernel_manager):
+        self.kernel_clients = {}
+        self.on_recv_funcs = {}
+        self.kernel_manager = kernel_manager
+        self.pool_index = 0
+        self.kernel_pool = []
+        self.kernel_semaphore = Semaphore(prespawn_count)
+
+        for _ in range(prespawn_count):
+            kernel_id = kernel_manager.start_kernel()
+            self.kernel_clients[kernel_id] = kernel_manager.get_kernel(kernel_id).client()
+            self.kernel_pool.append(kernel_id)
+            iopub = self.kernel_manager.connect_iopub(kernel_id)
+            iopub.on_recv(self.create_on_reply(kernel_id))
+
+    @gen.coroutine
+    def acquire(self):
```

Comment:
So you did end up having to acquire and release? Queuing up requests on a kernel didn't work?